File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
  def longestPalindrome(self, s: str) -> str:
    if len(s) == 1: return s
    preprocess = '^#' + '#'.join(s) + '#$'
    dp = [1] * len(preprocess)
    index = 1
    center = 0
    right = 0
    for i in range(1, len(preprocess)-3):
      if i < (right-1):
        dp[i] = min(right-i, dp[2*center-i])
      while preprocess[i-dp[i]] == preprocess[i+dp[i]]:
        dp[i] += 1
      if i + dp[i] > right:
        center = i
        right = center + dp[i]
      if dp[i] > dp[index]:
        index = i
    
    return s[(index-dp[index])//2:(index-dp[index])//2+dp[index]-1]
    

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cs
public class Solution {
  public string LongestPalindrome(string s) {
    string str = "^#" + String.Join('#', s as IEnumerable<char>) + "#$";
    int index = 0;
    int center = 0;
    int right = 0;
    int n = str.Length;
    int[] dp = Enumerable.Repeat<int>(1, n).ToArray();
    
    for (int i = 1; i < n - 2; i++) {
      if (i < right - 1) {
        dp[i] = Math.Min(right - i, dp[2*center - i]);
      }
      while (str[i-dp[i]] == str[i+dp[i]]) {
        dp[i] += 1;
      }
      if (i + dp[i] > right) {
        center = i;
        right = i + dp[i];
      }
      if (dp[i] > dp[index]) {
        index = i;
      }
    }
    return s.Substring((index - dp[index])/2, dp[index]-1);
  }
}